File: CSharp/Legacy89DiskKit.Infrastructure/DiskImage/Container/D88DiskContainer.cs
using Legacy89DiskKit.Domain.DiskImage.Interface.Container;
using Legacy89DiskKit.Domain.DiskImage.Model;
using Legacy89DiskKit.Domain.DiskImage.Exception;
using Legacy89DiskKit.Infrastructure.DiskImage.D88;
using System.Text;

namespace Legacy89DiskKit.Infrastructure.DiskImage.Container;

public class D88DiskContainer : IDiskContainer, IDisposable
{
    private string _filePath;
    private bool _isReadOnly;
    private byte[] _imageData = Array.Empty<byte>();
    private D88Header _header = new D88Header();
    private readonly Dictionary<(int, int, int), D88SectorData> _sectors;
    private bool _hasChanges = false;

    public string FilePath => _filePath;
    public bool IsReadOnly => _isReadOnly;
    public DiskType DiskType => _header.MediaType;

    public D88DiskContainer(string filePath, bool isReadOnly = false)
        : this(LoadDiskImage(filePath), isReadOnly, filePath)
    {
    }

    /// <summary>
    /// Initializes a D88 container from an in-memory disk image.
    /// </summary>
    public D88DiskContainer(byte[] imageData, bool isReadOnly = true, string filePath = "")
        : this(imageData, isReadOnly, filePath, false)
    {
    }

    private D88DiskContainer(byte[] imageData, bool isReadOnly, string filePath, bool skipClone)
    {
        _filePath = filePath;
        _isReadOnly = isReadOnly;
        _sectors = new Dictionary<(int, int, int), D88SectorData>();
        _imageData = skipClone ? imageData : (byte[])imageData.Clone();
        LoadFromBytes();
    }

    public static D88DiskContainer CreateNew(string filePath, DiskType diskType, string diskName = "", int? sectorsPerTrack = null, ushort? sectorSize = null)
    {
        var container = new D88DiskContainer();
        container._filePath = filePath;
        container._isReadOnly = false;
        container.CreateEmptyImage(diskType, diskName, sectorsPerTrack, sectorSize);
        container.SaveToFile();
        return container;
    }

    public static D88DiskContainer CreateNew(string filePath, DiskType diskType, string diskName, Func<int, int, (int sectors, ushort size, byte density)?> perTrackGeometry)
    {
        var container = new D88DiskContainer();
        container._filePath = filePath;
        container._isReadOnly = false;
        container.CreateEmptyImage(diskType, diskName, perTrackGeometry: perTrackGeometry);
        container.SaveToFile();
        return container;
    }

    /// <summary>
    /// Creates a new in-memory D88 container.
    /// </summary>
    public static D88DiskContainer CreateNewInMemory(string diskName, DiskType diskType, int? sectorsPerTrack = null, ushort? sectorSize = null)
    {
        var container = new D88DiskContainer();
        container._filePath = "";
        container._isReadOnly = false;
        container.CreateEmptyImage(diskType, diskName, sectorsPerTrack, sectorSize);
        container.BuildImageData();
        return new D88DiskContainer(container._imageData, false);
    }

    private D88DiskContainer()
    {
        _filePath = "";
        _isReadOnly = false;
        _sectors = new Dictionary<(int, int, int), D88SectorData>();
    }

    private static byte[] LoadDiskImage(string filePath)
    {
        if (filePath is null)
        {
            throw new ArgumentNullException(nameof(filePath));
        }

        if (!File.Exists(filePath))
        {
            throw new FileNotFoundException($"D88 file not found: {filePath}");
        }

        return File.ReadAllBytes(filePath);
    }

    private void LoadFromBytes()
    {
        try
        {
            if (_imageData.Length < 0x2b0)
            {
                throw new DiskImageException($"Invalid D88 file: too small ({_imageData.Length} bytes)");
            }

            _header = D88ImageParser.ParseHeader(_imageData);
            var parsedSectors = D88ImageParser.ParseSectors(_imageData, _header);
            _sectors.Clear();
            foreach (var entry in parsedSectors)
            {
                _sectors[entry.Key] = entry.Value;
            }
        }
        catch (Exception ex) when (ex is not DiskImageException)
        {
            throw new DiskImageException($"Error loading D88 file: {ex.Message}", ex);
        }
    }

    private static int GetMaxSectorsPerTrack(DiskType diskType)
    {
        return diskType switch
        {
            DiskType.TwoD => 16,
            DiskType.TwoDD => 16,
            DiskType.TwoHD => 26,
            _ => 26
        };
    }

    public byte[] ReadSector(int cylinder, int head, int sector)
    {
        if (!_sectors.TryGetValue((cylinder, head, sector), out var d88Sector))
            throw new DiskImageException($"Sector not found: C={cylinder}, H={head}, R={sector}");
        return d88Sector.Data;
    }

    public DiskContainerMetadata GetMetadata()
    {
        var metadata = D88ImageParser.CreateMetadata(_header, _sectors.Values);
        return metadata with { IsWriteProtected = metadata.IsWriteProtected || _isReadOnly };
    }

    public byte[] ReadSector(int cylinder, int head, int sector, bool allowCorrupted)
    {
        return ReadSector(cylinder, head, sector);
    }

    public void WriteSector(int cylinder, int head, int sector, byte[] data)
    {
        if (_isReadOnly) throw new DiskImageException("Disk image is read-only");
        if (!_sectors.TryGetValue((cylinder, head, sector), out var d88Sector))
            throw new DiskImageException($"Sector not found: C={cylinder}, H={head}, R={sector}");
            
        d88Sector.Data = (byte[])data.Clone();
        d88Sector.ActualSize = (ushort)data.Length;
        _hasChanges = true;
    }

    public bool SectorExists(int cylinder, int head, int sector)
    {
        return _sectors.ContainsKey((cylinder, head, sector));
    }

    public IEnumerable<SectorInfo> GetAllSectors()
    {
        return _sectors.Values.Select(s => new SectorInfo(
            s.Cylinder, s.Head, s.Sector, s.ActualSize, s.Deleted, s.Status != 0));
    }

    public void Save() => SaveAs(_filePath);

    public void SaveAs(string filePath)
    {
        if (string.IsNullOrEmpty(filePath))
            throw new DiskImageException("Cannot save a disk image without a file path");
        BuildImageData();
        File.WriteAllBytes(filePath, _imageData);
        _hasChanges = false;
        _filePath = filePath;
    }

    /// <summary>
    /// Returns the current D88 image bytes.
    /// </summary>
    public byte[] ToImageData()
    {
        if (_hasChanges || _imageData.Length == 0)
        {
            BuildImageData();
        }

        return (byte[])_imageData.Clone();
    }

    private void BuildImageData()
    {
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream);
        
        // Write Name (17 bytes)
        var nameBytes = new byte[17];
        var nameData = Encoding.ASCII.GetBytes(_header.ImageName);
        Array.Copy(nameData, nameBytes, Math.Min(nameData.Length, 17));
        writer.Write(nameBytes);
        
        writer.Write(new byte[9]); // reserved
        writer.Write((byte)(_header.WriteProtect ? 0x10 : 0x00));
        writer.Write((byte)_header.MediaType);
        
        long diskSizePos = writer.BaseStream.Position;
        writer.Write((uint)0); // Placeholder for DiskSize

        long trackOffsetsPos = writer.BaseStream.Position;
        var trackOffsets = new uint[164];
        for (int i = 0; i < 164; i++) writer.Write((uint)0); // Placeholder for track offsets

        var currentOffset = 0x2b0u;
        for (int track = 0; track < 164; track++)
        {
            var cylinder = track / 2;
            var head = track % 2;
            var trackSectors = _sectors.Values
                .Where(s => s.Cylinder == cylinder && s.Head == head)
                .OrderBy(s => s.Sector)
                .ToList();

            if (trackSectors.Any())
            {
                trackOffsets[track] = currentOffset;
                foreach (var sector in trackSectors)
                {
                    writer.Write(sector.Cylinder);
                    writer.Write(sector.Head);
                    writer.Write(sector.Sector);
                    writer.Write(sector.SectorSizeN);
                    writer.Write((ushort)trackSectors.Count);
                    writer.Write(sector.Density);
                    writer.Write((byte)(sector.Deleted ? 0x10 : 0x00));
                    writer.Write(sector.Status);
                    writer.Write(new byte[5]);
                    writer.Write((ushort)sector.Data.Length);
                    writer.Write(sector.Data);
                    currentOffset += 16u + (uint)sector.Data.Length;
                }
            }
        }
        
        // Finalize header
        uint totalSize = (uint)writer.BaseStream.Length;
        writer.BaseStream.Seek(diskSizePos, SeekOrigin.Begin);
        writer.Write(totalSize);
        
        writer.BaseStream.Seek(trackOffsetsPos, SeekOrigin.Begin);
        for (int i = 0; i < 164; i++) writer.Write(trackOffsets[i]);
        
        _imageData = stream.ToArray();
    }

    private void CreateEmptyImage(DiskType diskType, string diskName, int? sectorsPerTrackParam = null, ushort? sectorSizeParam = null, Func<int, int, (int sectors, ushort size, byte density)?>? perTrackGeometry = null)
    {
        _header = new D88Header { ImageName = diskName, MediaType = diskType };
        _sectors.Clear();

        int maxCylinders = diskType switch { DiskType.TwoHD => 77, DiskType.TwoDD => 80, _ => 40 };
        int maxHeads = 2;
        int defaultSpt = sectorsPerTrackParam ?? GetMaxSectorsPerTrack(diskType);
        ushort defaultSize = sectorSizeParam ?? (ushort)(diskType == DiskType.TwoHD ? 1024 : 256);
        byte defaultDensity = (byte)(diskType == DiskType.TwoHD ? 0x01 : 0x00);

        for (int c = 0; c < maxCylinders; c++)
        {
            for (int h = 0; h < maxHeads; h++)
            {
                var over = perTrackGeometry?.Invoke(c, h);
                int   spt     = over?.sectors ?? defaultSpt;
                ushort sz     = over?.size    ?? defaultSize;
                byte density  = over?.density ?? defaultDensity;
                byte sizeN    = sz switch { 256 => 1, 512 => 2, 1024 => 3, _ => 2 };

                for (int s = 1; s <= spt; s++)
                {
                    var d88Sector = new D88SectorData
                    {
                        Cylinder    = (byte)c,
                        Head        = (byte)h,
                        Sector      = (byte)s,
                        SectorSizeN = sizeN,
                        SectorCount = (ushort)spt,
                        Density     = density,
                        Deleted     = false,
                        Status      = 0,
                        ActualSize  = sz,
                        Data        = new byte[sz]
                    };
                    _sectors[(c, h, s)] = d88Sector;
                }
            }
        }
    }

    private void SaveToFile() => Save();

    public void Dispose()
    {
        if (_hasChanges && !IsReadOnly && !string.IsNullOrEmpty(_filePath))
        {
            try { Save(); } catch { }
        }
    }

}

File: Cpp/Legacy89DiskKit.Cpp/src/infrastructure/disk_image/d88_disk_container.cpp
#include "legacy89diskkit/cpp/infrastructure/disk_image/d88_disk_container.hpp"

#include "legacy89diskkit/cpp/d88_parser.hpp"

#include <algorithm>
#include <array>

namespace legacy89diskkit::cpp
{
namespace
{
std::uint8_t DetermineSectorSizeCode(const int size)
{
    switch (size)
    {
    case 128:
        return 0;
    case 256:
        return 1;
    case 512:
        return 2;
    case 1024:
        return 3;
    default:
        return 0;
    }
}
}

Result<D88DiskContainer> D88DiskContainer::OpenFromBuffer(const std::span<const std::uint8_t> image_data, const bool read_only)
{
    const auto header_result = D88Parser::ParseHeader(std::vector<std::uint8_t>(image_data.begin(), image_data.end()));
    if (!header_result.ok())
    {
        return Result<D88DiskContainer>::Failure(header_result.status().code, header_result.status().message);
    }

    const auto layout_result = D88Parser::ParseImage(std::vector<std::uint8_t>(image_data.begin(), image_data.end()));
    if (!layout_result.ok())
    {
        return Result<D88DiskContainer>::Failure(layout_result.status().code, layout_result.status().message);
    }

    auto metadata = layout_result.value().metadata;
    metadata.is_write_protected = metadata.is_write_protected || read_only;
    return Result<D88DiskContainer>::Success(D88DiskContainer(
        header_result.value().image_name,
        layout_result.value().sectors,
        std::move(metadata),
        read_only,
        {}));
}

Result<D88DiskContainer> D88DiskContainer::CreateNew(DiskType type, const std::string& name)
{
    const int cylinders = (type == DiskType::TwoHD) ? 77 : 40;
    const int heads = 2;
    const int sectors_per_track = (type == DiskType::TwoHD) ? 26 : 16;
    const int sector_size = (type == DiskType::TwoHD) ? 1024 : 256;

    std::vector<SectorDataBlock> sectors;
    sectors.reserve(static_cast<std::size_t>(cylinders * heads * sectors_per_track));

    for (int c = 0; c < cylinders; ++c)
    {
        for (int h = 0; h < heads; ++h)
        {
            for (int s = 1; s <= sectors_per_track; ++s)
            {
                sectors.push_back(SectorDataBlock{
                    SectorInfo{c, h, s, sector_size, false, false},
                    std::vector<std::uint8_t>(static_cast<std::size_t>(sector_size), 0x00)});
            }
        }
    }

    DiskContainerMetadata metadata{};
    metadata.image_format = "d88-sector-container";
    metadata.disk_type = type;
    metadata.geometry = DiskGeometryInfo{cylinders, heads, sectors_per_track, sector_size};
    metadata.is_write_protected = false;

    return Result<D88DiskContainer>::Success(D88DiskContainer(
        name,
        std::move(sectors),
        std::move(metadata),
        false,
        {}));
}

D88DiskContainer::D88DiskContainer(
    std::string image_name,
    std::vector<SectorDataBlock> sectors,
    DiskContainerMetadata metadata,
    const bool read_only,
    std::string file_path)
    : image_name_(std::move(image_name)),
      sectors_(std::move(sectors)),
      metadata_(std::move(metadata)),
      read_only_(read_only),
      file_path_(std::move(file_path))
{
}

const std::string& D88DiskContainer::FilePath() const
{
    return file_path_;
}

bool D88DiskContainer::IsReadOnly() const
{
    return read_only_;
}

DiskType D88DiskContainer::DiskTypeValue() const
{
    return metadata_.disk_type;
}

DiskContainerMetadata D88DiskContainer::GetMetadata() const
{
    return metadata_;
}

Result<std::vector<std::uint8_t>> D88DiskContainer::ReadSector(const int cylinder, const int head, const int sector) const
{
    const auto it = std::find_if(
        sectors_.begin(),
        sectors_.end(),
        [&](const auto& block)
        {
            return block.sector.cylinder == cylinder &&
                   block.sector.head == head &&
                   block.sector.sector == sector;
        });
    if (it == sectors_.end())
    {
        return Result<std::vector<std::uint8_t>>::Failure(StatusCode::OutOfRange, "Sector not found.");
    }

    return Result<std::vector<std::uint8_t>>::Success(it->data);
}

Status D88DiskContainer::WriteSector(const int cylinder, const int head, const int sector, const std::span<const std::uint8_t> data)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "D88 container is read-only."};
    }

    const auto it = std::find_if(
        sectors_.begin(),
        sectors_.end(),
        [&](const auto& block)
        {
            return block.sector.cylinder == cylinder &&
                   block.sector.head == head &&
                   block.sector.sector == sector;
        });
    if (it == sectors_.end())
    {
        return {StatusCode::OutOfRange, "Sector not found."};
    }

    it->sector.size = static_cast<int>(data.size());
    it->data.assign(data.begin(), data.end());
    return Status::OkStatus();
}

bool D88DiskContainer::SectorExists(const int cylinder, const int head, const int sector) const
{
    return std::any_of(
        sectors_.begin(),
        sectors_.end(),
        [&](const auto& block)
        {
            return block.sector.cylinder == cylinder &&
                   block.sector.head == head &&
                   block.sector.sector == sector;
        });
}

std::vector<SectorInfo> D88DiskContainer::GetAllSectors() const
{
    std::vector<SectorInfo> sectors;
    sectors.reserve(sectors_.size());
    for (const auto& block : sectors_)
    {
        sectors.push_back(block.sector);
    }
    return sectors;
}

std::vector<std::uint8_t> D88DiskContainer::ToImageData() const
{
    std::vector<std::uint8_t> image(0x2b0, 0x00);
    const auto copy_length = std::min<std::size_t>(image_name_.size(), 17);
    std::copy_n(image_name_.begin(), copy_length, image.begin());
    image[0x1a] = metadata_.is_write_protected ? 0x10 : 0x00;
    image[0x1b] = static_cast<std::uint8_t>(metadata_.disk_type);

    std::array<std::uint32_t, 164> track_offsets{};
    std::uint32_t current_offset = 0x2b0;
    std::size_t track_offset_position = 0x20;

    for (int cylinder = 0; cylinder < metadata_.geometry.cylinders; ++cylinder)
    {
        for (int head = 0; head < metadata_.geometry.heads; ++head)
        {
            std::vector<const SectorDataBlock*> track_sectors;
            for (const auto& sector : sectors_)
            {
                if (sector.sector.cylinder == cylinder && sector.sector.head == head)
                {
                    track_sectors.push_back(&sector);
                }
            }

            if (track_sectors.empty())
            {
                continue;
            }

            const auto track_index = static_cast<std::size_t>(cylinder * 2 + head);
            if (track_index < track_offsets.size())
            {
                track_offsets[track_index] = current_offset;
            }

            std::sort(
                track_sectors.begin(),
                track_sectors.end(),
                [](const auto* left, const auto* right)
                {
                    return left->sector.sector < right->sector.sector;
                });

            for (const auto* sector : track_sectors)
            {
                image.push_back(static_cast<std::uint8_t>(sector->sector.cylinder));
                image.push_back(static_cast<std::uint8_t>(sector->sector.head));
                image.push_back(static_cast<std::uint8_t>(sector->sector.sector));
                image.push_back(DetermineSectorSizeCode(sector->sector.size));
                image.push_back(static_cast<std::uint8_t>(track_sectors.size() & 0xff));
                image.push_back(static_cast<std::uint8_t>((track_sectors.size() >> 8) & 0xff));
                image.push_back(metadata_.disk_type == DiskType::TwoHD ? 0x01 : 0x00);
                image.push_back(sector->sector.is_deleted ? 0x10 : 0x00);
                image.push_back(sector->sector.has_error ? 0x01 : 0x00);
                image.insert(image.end(), 5, 0x00);
                image.push_back(static_cast<std::uint8_t>(sector->data.size() & 0xff));
                image.push_back(static_cast<std::uint8_t>((sector->data.size() >> 8) & 0xff));
                image.insert(image.end(), sector->data.begin(), sector->data.end());
                current_offset += static_cast<std::uint32_t>(16 + sector->data.size());
            }
        }
    }

    const auto disk_size = static_cast<std::uint32_t>(image.size());
    image[0x1c] = static_cast<std::uint8_t>(disk_size & 0xff);
    image[0x1d] = static_cast<std::uint8_t>((disk_size >> 8) & 0xff);
    image[0x1e] = static_cast<std::uint8_t>((disk_size >> 16) & 0xff);
    image[0x1f] = static_cast<std::uint8_t>((disk_size >> 24) & 0xff);

    for (const auto track_offset : track_offsets)
    {
        image[track_offset_position + 0] = static_cast<std::uint8_t>(track_offset & 0xff);
        image[track_offset_position + 1] = static_cast<std::uint8_t>((track_offset >> 8) & 0xff);
        image[track_offset_position + 2] = static_cast<std::uint8_t>((track_offset >> 16) & 0xff);
        image[track_offset_position + 3] = static_cast<std::uint8_t>((track_offset >> 24) & 0xff);
        track_offset_position += 4;
    }

    return image;
}
}
